### DISCUSSION TITLE
Shouldn't `eglot-workspace-configuration` be safe as a function symbol when setting its directory-local value?

### DISCUSSION THREAD

**Discussion by galeo** [joaotavora/eglot#1392] [General]
As its docstring describes, the directory-local value of `eglot-workspace-configuration` can be set to a unary function. However, the code does not correspond to this description.

```elisp
;;;###autoload
(put 'eglot-workspace-configuration 'safe-local-variable #'listp)
```

Is this intentional or does it need to be adjusted?

**Comment by joaotavora** [comment]
Well generally, things that are functions aren't really safe when used as file or directory local variables, because functions have side-effects, at least how _I_ understand what "safe" means in this context.  But I'm definitely not an expert.  But just imagine I give you a project to work on with a directory local definition of `e-w-c` with a function `(lambda (&rest _ignore) (spawn-malware))` and tell you to run `M-x eglot` on it.

**Reply by joaotavora** [reply]
So if you want to run that risk (you might want, to, after considering things), just add a more permissive safety spec to your config.

**Reply by galeo** [reply]
Thanks. However, providing a unified basic configuration for multiple projects of the same lsp server through `eglot-workspace-configuration` requires many unnecessary copy operations. For the same LSP server, there are many same basic configurations.
Just like `eglot-server-programs`, wouldn't it be better to set a base configuration for different lsp servers and allow overriding the base configuration in a specific project?

**Reply by galeo** [reply]
Additionally, I took a look at #1279 and the [User-spcific configuration section](https://www.gnu.org/software/emacs/manual/html_node/eglot/User_002dspecific-configuration.html) in the documentation. There are problems when using `:initializationOptions` or setting `eglot-workspace-configuration` via `setq-default`. The former is not supported by some lsp servers, and the latter may cause errors in some lsp servers.

**Reply by galeo** [reply]
My current approach is to set the directory-local value of `eglot-workspace-configuration` to a function to simplify the copy operation as much as possible, but it is still inconvenient to use. So would you like to consider providing a setting item probably named `workspace-base-configuration` to uniformly set the basic configuration of a certain LSP server?

**Reply by joaotavora** [reply]
This discussion started about the dir/file-local safeness of a value, then it pivoted to something else about "unified configurations" I can't understand.

**Reply by galeo** [reply]
Sorry. It's indeed a bit off topic. The "unified configuration" here means that the same basic configuration applies to multiple projects used by the same lsp server.

**Reply by joaotavora** [reply]
> multiple projects

multiple === all?

**Reply by joaotavora** [reply]
Because if it is `multiple === all`, you'd be better served by the methods described here: https://joaotavora.github.io/eglot/#User_002dspecific-configuration.  Actually start by reading: https://joaotavora.github.io/eglot/#Advanced-server-configuration-1.

There are three distinct methods described available to you: server-specific RC files, command-line options and LSP `initializationOptions`.  Do investigate if your server supports any of them.  


Another entirely-different thing is project-specific configuration https://joaotavora.github.io/eglot/#Project_002dspecific-configuration-1. Many times, many users are trying to (ab)use `eglot-workspace-configuration` to achieve user-specific per-server configuration, when in effect this "workspace configuration" was intended -- by LSP, not by Eglot -- as a means of project-specific configuration.  

Nevertheless this abuse can be achieved by Eglot, just search for `setq-default` within that section

**Reply by galeo** [reply]
Thanks for the summary. The `setq-default` way will cause errors in some lsp servers, such as [zls](https://github.com/zigtools/zls), which will cause a segmentation fault when it receives additional setting items.

**Reply by joaotavora** [reply]
> Thanks for the summary. The setq-default way will cause errors in some lsp servers, such as [zls](https://github.com/zigtools/zls), which will cause a segmentation fault when it receives additional setting items.

I didn't know that.  It could be that this is a bug in the `zls` server, but I'm not sure what the standard says about that.

Anyway, the `setq-default` way, can still be done, right?  just:

```
(setq-default eglot-workspace-configuration #'function-that-doesnt-send-items-other-the-ones-server-wants)
```

**Reply by galeo** [reply]
Although this method is somewhat of a hack, it is helpful. Thanks.

**Reply by joaotavora** [reply]
> Although this method is somewhat of a hack, it is helpful. Thanks.

Using a mechanism for project-specific configuration to accomplish user-specific, per-server configuration _is_ a hack.
